clang/lib/Sema/SemaTemplateInstantiate.cpp

```diff
@@ -285,7 +286,8 @@ Response HandleFunction(Sema &SemaRef, const FunctionDecl *Function,
     // If this function was instantiated from a specialized member that is
     // a function template, we're done.
     assert(Function->getPrimaryTemplate() && "No function template?");
-    if (Function->getPrimaryTemplate()->isMemberSpecialization())
+    if (!ForDefaultArgumentSubstitution &&
+        Function->getPrimaryTemplate()->isMemberSpecialization())
```

Comment:
An alternative is to use `(ForConstraintInstantiation || !RelativeToPrimary)` so we could get rid of the usages for concepts where it expects an end after a member specialization. But I feel like that makes the semantics of the flags vague, so ideas are welcome here.

Reply:
looks like https://github.com/llvm/llvm-project/pull/102267 maybe ?

Reply:
Nope, I applied that patch and the problem persists.
I stepped in the logic a little, and it seemed we had bailed out too early before the specifier substitution.

Reply:
Hum. But yes, it should be valid

Reply:
Filed https://github.com/llvm/llvm-project/issues/105462 and let's focus on this PR again.
Despite that, I now realize using `(ForConstraintInstantiation || !RelativeToPrimary)` is probably unfeasible. The noexcept specifier would expect an end after such template specialization because it doesn't *copy* the expression from the primary declaration, which makes it impossible to refer to primary template parameters.

How about proceeding with the new bool flag and refactoring the function to use enums in the next next (the refactoring never ends ;) PR? As that would entail a lot of changes from other files, we can make the review easier

Reply:
Works for me

Reply:
Thanks!